lib/vector-tap/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+                        self.component_kind_label,
+                        component_kind.green(),
+                        self.component_type_label,
+                        component_type.green()
+                    )
+                        .into()
+                }
+                TapEncodingFormat::Logfmt => format!(
+                    "{}={} {}={} {}={} {}",
+                    self.component_id_label,
+                    component_id.green(),
+                    self.component_kind_label,
+                    component_kind.green(),
+                    self.component_type_label,
+                    component_type.green(),
+                    event
+                )
+                    .into(),
+            }
+        } else {
+            event.into()
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum OutputChannel {
+    Stdout(EventFormatter),
+    AsyncChannel(tokio_mpsc::Sender<Vec<GraphQLTapOutputEvents>>),
+}
+
+/// Error type for DNS message parsing
+#[derive(Debug, Snafu)]
+pub enum TapExecutorError {
+    #[snafu(display("[tap] Couldn't connect to API via WebSockets"))]
+    ConnectionFailure,
+    GraphQLError,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn exec_tap(
+    url: &Url,
+    interval: i64,
+    limit: i64,
+    duration_ms: Option<u64>,
+    input_patterns: Vec<String>,
+    output_patterns: Vec<String>,
+    format: TapEncodingFormat,
+    formatter: &EventFormatter,
+    quiet: bool,
+) -> Result<(), TapExecutorError> {
+    let subscription_client = match connect_subscription_client((*url).clone()).await {
+        Ok(c) => c,
+        Err(e) => {
+            #[allow(clippy::print_stderr)]
+            {
+                eprintln!("[tap] Couldn't connect to API via WebSockets: {}", e);
+            }
+            return Err(TapExecutorError::ConnectionFailure);
+        }
+    };
+
+    tokio::pin! {
+        let stream = subscription_client.output_events_by_component_id_patterns_subscription(
+            output_patterns.clone(),
+            input_patterns.clone(),
+            format,
+            limit,
+            interval,
+        );
+    }
+
+    let start_time = Instant::now();
+    let stream_duration =
+        duration_ms
+            .map(Duration::from_millis)
+            .unwrap_or(Duration::MAX);
+
+    // Loop over the returned results, printing out tap events.
+    #[allow(clippy::print_stdout)]
+    #[allow(clippy::print_stderr)]
+    loop {
+        let time_elapsed = start_time.elapsed();
+        if time_elapsed >= stream_duration {
+            return Ok(());
+        }
+
+        let message = timeout(stream_duration - time_elapsed, stream.next()).await;
+        match message {
+            Ok(Some(Some(res))) => {
+                if let Some(d) = res.data {
+                    for tap_event in d.output_events_by_component_id_patterns.iter() {
+                        match tap_event {
+                            GraphQLTapOutputEvents::Log(ev) => {
```

Comment:
Nit: It would be nice to move this to each own function. But if this causes conflict with the next PR, then no need to change it in this one.

Reply:
Moving to separate function in next PR